tests/lang.sh

```diff
@@ -24,59 +60,89 @@ fail=0
+            2> >(sed "s~$(pwd)~/pwd~g" > "lang/$i.err")
+    then
         echo "FAIL: $i should parse"
         fail=1
+    else
+        diffAndAccept "$i" out exp
+        diffAndAccept "$i" err err.exp
     fi
 done
 
 for i in lang/eval-fail-*.nix; do
     echo "evaluating $i (should fail)";
-    i=$(basename $i .nix)
-    if ! expect 1 nix-instantiate --eval lang/$i.nix; then
+    i=$(basename "$i" .nix)
+    if
+        ! expectStderr 1 nix-instantiate --show-trace "lang/$i.nix" \
+            | sed "s~$(pwd)~/pwd~g" > "lang/$i.err"
+    then
         echo "FAIL: $i shouldn't evaluate"
         fail=1
+    else
+        diffAndAccept "$i" err err.exp
     fi
 done
 
 for i in lang/eval-okay-*.nix; do
     echo "evaluating $i (should succeed)";
-    i=$(basename $i .nix)
+    i=$(basename "$i" .nix)
 
-    if test -e lang/$i.exp; then
-        flags=
-        if test -e lang/$i.flags; then
-            flags=$(cat lang/$i.flags)
-        fi
-        if ! expect 0 env NIX_PATH=lang/dir3:lang/dir4 HOME=/fake-home nix-instantiate $flags --eval --strict lang/$i.nix > lang/$i.out; then
+    if test -e "lang/$i.exp.xml"; then
+        if ! expect 0 nix-instantiate --eval --xml --no-location --strict \
+                "lang/$i.nix" > "lang/$i.out.xml"; then
             echo "FAIL: $i should evaluate"
             fail=1
-        elif ! diff <(< lang/$i.out sed -e "s|$(pwd)|/pwd|g") lang/$i.exp; then
-            echo "FAIL: evaluation result of $i not as expected"
-            fail=1
+        else
+            diffAndAccept "$i" out.xml exp.xml
+        fi
+    elif test ! -e "lang/$i.exp-disabled"; then
+        declare -a flags=()
+        if test -e "lang/$i.flags"; then
+            read -r -a flags < "lang/$i.flags"
         fi
-    fi
 
-    if test -e lang/$i.exp.xml; then
-        if ! expect 0 nix-instantiate --eval --xml --no-location --strict \
-                lang/$i.nix > lang/$i.out.xml; then
+        if
+            ! expect 0 env \
+                NIX_PATH=lang/dir3:lang/dir4 \
+                HOME=/fake-home \
+                nix-instantiate "${flags[@]}" --eval --strict "lang/$i.nix" \
+                1> "lang/$i.out" \
+                2> "lang/$i.err"
+        then
             echo "FAIL: $i should evaluate"
             fail=1
-        elif ! cmp -s lang/$i.out.xml lang/$i.exp.xml; then
-            echo "FAIL: XML evaluation result of $i not as expected"
-            fail=1
+        else
+            sed -i "s~$(pwd)~/pwd~g" "lang/$i.out" "lang/$i.err"
+            diffAndAccept "$i" out exp
+            diffAndAccept "$i" err err.exp
         fi
     fi
 done
 
+if (( "$fail" )); then
+    echo ''
+    echo 'You can rerun this test with:'
+    echo ''
+    echo '    _NIX_TEST_ACCEPT=1 make tests/lang.sh.test'
+    echo ''
+    echo 'to regenerate the files containing the expected output,'
+    echo 'and then view the git diff to decide whether a change is'
+    echo 'good/intenation or bad/accidental.'
```

Comment:
Typo

```suggestion
    echo 'good/intentional or bad/accidental.'
```

Not accidental as in complexity. Although that might be an uncommon interpretation, I don't think accidental adds much here.

```suggestion
    echo 'good/intentional or bad.'
```

```suggestion
    echo 'good/intentional or bad.'
    echo 'If the diff contains arbitrary or impure information,'
    echo 'please improve the normalization that the test applies to the output.'
```

More complete or too much hand holding?